### DISCUSSION TITLE
How is the execution configuration different in Warp than CUDA?

### DISCUSSION THREAD

**Discussion by AkeelMedina22** [NVIDIA/warp#78] [General]
Hello everyone, 

I'm a beginner to CUDA and Warp, and I was wondering how concepts such as Blocks and Threads work in Warp. For example, if I wanted to do a simple dot product, in CUDA I would use several blocks with an ideal number of threads as per the occupancy calculator. Would launching a multidimensional grid be appropriate to mimic this? Or does this happen automatically with a 1-dimensional execution configuration?

**Comment by mmacklin** [comment]
Hi @AkeelMedina22 , Warp takes care of decomposing kernel thread grids into blocks for you - we currently use a fixed block dimension of 256 threads (regardless of single / multi-dimension launches).

If we were to expose a way set the block dimensions inside `wp.launch()` would that be useful for you?

**Reply by nishadgothoskar** [reply]
if the block dimension is always 256, how does the number of blocks get selected? I see the generated cuda code is set up using a for loop, following [Grid-Stride Loops](https://developer.nvidia.com/blog/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/), but I haven't been able to identify how the number of blocks is selected.

**Reply by shi-eric** [reply]
In Warp 1.5.0, `block_dim` will be exposed as an argument to `wp.launch()` to control the number of threads per block. 

The number of blocks gets selected in the following code: https://github.com/NVIDIA/warp/blob/1bab08df396ae891e9bd1b91af49ba61f7e6a5e2/warp/native/warp.cu#L3150-L3202

We compute the number of blocks that would be required using to process one element of the Warp launch grid per thread. If this results in too many blocks than supported by the CUDA device, then we use grid-stride loops to process more than one element per thread.
